kernel: add new cifsd module and cifsd-tools

Andy2244 opened the pull request:
Maintainer: me
Compile tested: arm/mips (master)
Run tested: arm (master)

Description:

* adds cifs/smb kernel server module cifsd
* adds userspace tools (cifsd, cifsadmin)

This is a new, simple smb2/3 compatible fileserver. The project is still relatively new, but worked fine for me as simple to setup fileshare. The devs plan to get this merged into upstream eventually, but may take a while.

The advantages over samba3/4 is size and potentially speed. I could not really test this, since my router is too powerfully, so i always get max samba network speeds. Yet i did observe lower cpu usage with this server.

I tested video, music streaming which worked fine. I also tested running [portableapps.com](https://portableapps.com/) from a mounted share, including Firefox Portable, which also works now.

In general the devs try to reuse samba config options, so that the `smb.conf` file looks very similar. Some  advanced options you may use via samba are not yet implemented, like:
```
unix charset
smb encrypt
fstype
dos filemode
inherit owner
```

PS: The firmware size (squashfs-sysupgrade.bin) for mvebu (arm) increases by 860kb with it build, where 135kb is for the cifsd+tools and the remainder is for extra dependencies, notably glib2.

Andy2244 committed:
kernel: add new cifsd module and cifsd-tools

* adds cifs/smb kernel server module cifsd
* adds userspace tools (cifsd, cifsadmin)

Signed-off-by: Andy Walsh <andy.walsh44+github@gmail.com>

Andy2244 commented:
I think it works "reasonably well" in the context of someone who cant/wont use samba4, because of size and memory limitation and wants better speeds than samba3 or needs smbv3 support.

If you use my guest config example, just change the path and mount the location as network drive most common scenarios work fine. Even samba3 has problems running binarys directly from a share, because of locking and bad mapped memory region support.

PS: I mean we can keep it in snapshots and i will retest the next batch of dev merges. I was thinking having a alternative to samba4 is better, even if its not fully stable yet.

neheb commented:
111k + 32k. Not bad.

I just tried this on a ramips NAS style device. It seems to work reasonably well. Still the same Unicode bugs.

Note that something like this probably needs to replace samba3 as Windows 10 and modern Linux has obsoleted CIFSv1.

Andy2244 commented:
> 111k + 32k. Not bad.

Yeah but keep in mind that the tools need glib2, which is rarely used by packages, so its around 800kb total, still way ahead of samba4. 
Yes i also saw the potential to have something simple/small similar to the nfs-kernel-server.

PS: What "Unicode bugs" you got? Just want to check if its the same i got and have issues open for.

neheb commented:
Was copying my Music to a share. Got this error.

```
rsync: mkstemp "/home/mangix/test/AccurateRip/Pop/2NE1/2nd Mini Album/.01 - 내가 제일 잘 나가 (I Am the Best).flac.YFQzKt" failed: No such file or directory (2)
```

and many like it. With Samba 3, the file would transfer with a totally messed up ASCII filename. With cifsd, it errors. Not a big deal since I can use normal rsync to transfer them afterwards.

The reason I rsync to a Samba share is because of speed (4MB/s vs. 22MB/s). AES is slow on mt7621.

Andy2244 commented:
Ah ok, did not test this unicode scenario. This should probably work, since the devs are Korean and should have personal interest in getting this to work properly. I will add this to the issues, is there any more error output in the log?

neheb commented:
Hahaha I didn't immediately realize that example was Korean. I also get this spam in dmesg

```
kcifsd: check_invalid_char:143: found invalid character : 0x3f
```

Not sure if related.

Andy2244 commented:
> Mounting the share with vers=2.0 or 2.1 turns encryption off. 3 and above enable it.

Is there a unencrypted version for smb3 in the specs? If so i can ask for a option to disable encryption, since that's a valid use case for us.

neheb commented:
No idea. I have not tested Samba 4 with encryption disabled.

bobafetthotmail commented:
> > Mounting the share with vers=2.0 or 2.1 turns encryption off. 3 and above enable it.
> 
> Is there a unencrypted version for smb3 in the specs? If so i can ask for a option to disable encryption, since that's a valid use case for us.

Afaik yes, you can disable it in Windows server settings
https://blogs.msdn.microsoft.com/openspecification/2017/05/26/smb-2-and-smb-3-security-in-windows-10-the-anatomy-of-signing-and-cryptographic-keys/

It seems that there is nothing in the client that will refuse a connection if your server does not enable encryption, while there is a server-side option that will refuse connection to clients not supporting encryption.

Andy2244 commented:
* updated to latest git
* several of my reported issues have been fixed, including the explorer (windows) browse problems

Andy2244 commented:
* update to latest git
* remove cifsdpwd.db file, default dummy user
* use 8s ipc timout, to ensure service is stoped/closed fully
* fixes win10 access errors
* fix share properties page segfault

neheb commented:
May want to pass -DCONFIG_CIFSD_ACL to the cifsd module. mounting with mount.cifs -o cifsacl will fail.

Andy2244 commented:
* update to latest git
* build kmod with CONFIG_CIFSD_ACL

Andy2244 commented:
* update to latest git
* fixes unicode issues (inaccessible files, folders due to illegal chars)
* fixes issues regarding the ability to run binary's from share
* moved tools package to Filesystem (similar to nfsd)

PS: This update fixed two (namjaejeon/cifsd#175, namjaejeon/cifsd-tools#68) of the last big issues we had with this package and the only outstanding issues is related to more complex locking aka correctly running firefox portable directly from a network share.
I think we can wait if the locking issues can be fixed in the next couple of days, but from my point of view i'm gtg. What do you think @neheb ?

neheb commented:
CONFIG_NLS_DEFAULT could also get set to utf8 anyway.

I’ve been testing this since the beginning. Most of my problems have been solved.

Making encryption optional would also be nice. I currently work around this by setting server max protocol to 2.

namjaejeon commented:
@neheb no need to set utf8 in CONFIG_NLS_DEFAULT with the latest cifsd master. I add the patch(https://github.com/namjaejeon/cifsd/commit/5bef5624847cc1ff582bad7639c9f19095f7f437) to fix it. CONFIG_NLS_DEFAULT could be get set to old one(iso8859-1). and I add new patch(https://github.com/namjaejeon/cifsd/commit/cb0935bc2660f178a9ab005ebd92c248a518200e) to turn encryption off. Okay, I will add the parameter in smb.conf to control enable/disable.

neheb commented:
Well, yeah. For size reasons, it might make sense to default it to utf8 anyways. I wonder if there's anything besides cifsd that really uses NLS in OpenWrt though...

Disabling encryption by default is fine with me. If people really want encrypted file transfer they can use rsync or scp. It's faster anyway.

namjaejeon commented:
> I wonder if there's anything besides cifsd that really uses NLS in OpenWrt though...
It is no issue. The use of utf8 by cifsd is independent and does not affect other modules.

> Disabling encryption by default is fine with me. If people really want encrypted file transfer they can use rsync or scp. It's faster anyway.
Okay:) Thanks for your comment!

hauke commented on package/network/services/cifsd-tools/files/cifsd.init:
```diff
@@ -0,0 +1,56 @@
+	if (lsmod | grep cifsd &>/dev/null); then
+		rmmod cifsd &>/dev/null
+	fi
+	modprobe cifsd 2>/dev/null
```
Why is it needed to remove the kernel module before this gets started?

Andy2244 replied:
Some setting may not be fully reflected without a full remove, so i just added a extra try before start.

hauke commented on package/kernel/cifsd/Makefile:
```diff
@@ -0,0 +1,57 @@
+		M="$(PKG_BUILD_DIR)" \
+		CFLAGS="$(TARGET_CFLAGS)" \
+		CONFIG_CIFS_SERVER=m \
+		modules
```
Please use KERNEL_MAKE_FLAGS as a base, it already provides many of your manually defines.

Andy2244 replied:
ok will do could not find a good example, documentation for this.

Andy2244 replied:
done

hauke commented on package/kernel/cifsd/Makefile:
```diff
@@ -0,0 +1,57 @@
+		+kmod-crypto-ccm
+	KCONFIG:= \
+		CONFIG_KEYS=y \
+		CONFIG_CRYPTO_ARC4=y
```
CONFIG_CRYPTO_ARC4 is already compiled in in OpenWrt.

Andy2244 replied:
Just to ensure it stays this way if something changes or was locally changed. It was not clear to me how those cases are handled, since i assumed if you compile a custom kernel this setting always ensures you get a compatible one if you select this package?

diizzyy replied:
https://github.com/openwrt/openwrt/blob/master/target/linux/generic/config-4.14#L880
https://github.com/openwrt/openwrt/blob/master/target/linux/generic/config-4.19#L920
Unless you deliberately remove these lines it'll always be included

Andy2244 replied:
Ok i'm still confused, maybe you explain how things should work.

If you use `KCONFIG:=` and the package is selected than you always end up with a compatible kernel. So for this reason i don't understand why we would remove the options from the package? 

I understand that we use the `generic/config` template to ensure that those packages can be installed and work without needing a new kernel, but to me that's just a extra specific option we choose.

Is there any harm by including the needed options via `KCONFIG:=`? I find this more clear since it documents what options a module actually needs.

diizzyy replied:
I think the issue is that opkg doesn't know of kernel config depends which means that you can technically install a package which wont run on in the end unless whaterver you need is provided in a module.

hauke commented on package/kernel/cifsd/Makefile:
```diff
@@ -0,0 +1,57 @@
+		+kmod-crypto-aead \
+		+kmod-crypto-ccm
+	KCONFIG:= \
+		CONFIG_KEYS=y \
```
Why you add CONFIG_KEYS here it will be activated ijn all kernel images, please measure how much the size of a common kernel increases with this option set.

Andy2244 replied:
It was a requirement for cifsd, so i added it here to ensure the kernel is compatible if this package is selected.

Andy2244 replied:
has no impact on size

diizzyy replied:
Please make a PR/patch for both 4.14 and 4.19
I don't see any difference in size either on mvebu (arm)

Andy2244 replied:
What you mean i already changed this in this PR for 4.14 and 4.19 ?
Should `CONFIG_KEYS=y` be its own PR?

diizzyy replied:
Perferably, at least a separate commit.

Andy2244 replied:
done

hauke commented on package/kernel/cifsd/Makefile:
```diff
@@ -0,0 +1,57 @@
+define KernelPackage/fs-cifsd
+	SUBMENU:=Filesystems
+	TITLE:=CIFS/SMB kernel server support
+	FILES:=$(PKG_BUILD_DIR)/cifsd.$(LINUX_KMOD_SUFFIX)
```
LINUX_KMOD_SUFFIX was removed, please use .ko directly.

Andy2244 replied:
done

hauke commented on package/network/services/cifsd-tools/Makefile:
```diff
@@ -0,0 +1,53 @@
+  CATEGORY:=Network
+  SUBMENU:=Filesystem
+  TITLE:=Kernel CIFS/SMB server support and userspace tools
+  DEPENDS:=+kmod-fs-cifsd +glib2 +libnl-core +libnl-genl $(ICONV_DEPENDS)
```
is it possible to use libnl-tiny here? you probably have to patch the makefile to make it link against libnl-tiny.

Andy2244 replied:
Need to test.

Andy2244 replied:
The libnl-tiny headers are incompatible and result in compile errors.

hauke commented on package/network/services/cifsd-tools/files/cifsd.init:
```diff
@@ -0,0 +1,56 @@
+	fi
+
+	logger -t 'cifsd' "Starting CIFS/SMB userspace service."
+	/usr/bin/env LANG=en_US.UTF-8 /usr/sbin/cifsd
```
It would be nice if you could use procd to manage this daemon, procd also takes crase to restart it if it crashes and that it gets shut down later on.

Andy2244 replied:
See my comment at the ini file, atm its incompatible with procd and i would get restarts. It seems the way the 2 helper threads operate causes some problems with procd.

ynezz replied:
> See my comment at the ini file, atm its incompatible with procd and i would get restarts. It seems the way the 2 helper threads operate causes some problems with procd.

@Andy2244 Would be nice to get this procd issue fixed, instead of this workarounds. Could you please report this procd issue to `openwrt-devel@lists.openwrt.org` or at http://bugs.openwrt.org ?

namjaejeon commented:
@neheb I added the patch (https://github.com/namjaejeon/cifsd/commit/9b9f145c330532b4404f433346870e7e655e2ffb) to make signing is defautly disable with SMB3. You don't need to downgrade SMB version to avoid signing and encryption overhead with the latest cifsd.

neheb commented:
Signing is just for authentication, right? I just tested on Windows with the previous version and was getting ~50MB/s (probably the limit for this RAID array) with ~40% sys usage according to top. Better than the 5 that I was getting with encryption.

namjaejeon commented:
No, If server signing is enable, all messages (such as read/write) are signed as well as authentication. and It was causing performance down with SMB3 before. Yep, It is in my pocket. thinking check it after closing urgent issues and requests. Anyway. You can use cifsd with SMB3 now:)

Andy2244 commented:
* updated to latest git
* fixes locking errors (Firefox Portable)
* adds "create mask", "directory mask" options

PS: We have a new blocking issue atm: namjaejeon/cifsd/issues/188

Andy2244 commented:
* update to latest git (2019-02-25)
* fixes namjaejeon/cifsd/issues/188

PS: @hauke so what specific changes are still needed? We don't have any blocking issues anymore for the package itself.

nbd168 commented:
I think we need to either find a way to modularize the keys support, or disable cifsd by default. Otherwise this will likely add a significant amount of extra kernel bloat for all targets.

diizzyy commented:
@nbd168 
No difference in binary size far as I can tell, haven't looked at memory usage however

nbd168 commented:
Are you sure? There's a lot of code under security/keys which gets pulled in when CONFIG_KEYS is set to y.

diizzyy commented:
Yes, unless ```ls -l "build_dir/target-arm_cortex-a9+neon_musl_eabi/linux-mvebu_cortexa9/vmlinux"*``` is the wrong place to look at. At least on mvebu it's the same
@Andy2244 also observed this: https://github.com/openwrt/openwrt/pull/1788#discussion_r258066742

neheb commented:
> Yes, unless `ls -l "build_dir/target-arm_cortex-a9+neon_musl_eabi/linux-mvebu_cortexa9/vmlinux"*` is the wrong place to look at. At least on mvebu it's the same
> @Andy2244 also observed this: [#1788 (comment)](https://github.com/openwrt/openwrt/pull/1788#discussion_r258066742)

Check build_dir/target-arm_cortex-a9+neon_musl_eabi/linux-mvebu_cortexa9/.config to verify that CONFIG_KEYS is indeed not set.

diizzyy commented:
Hmm....
Looks like you need more stuff than just
```
CONFIG_ASSOCIATIVE_ARRAY=y
CONFIG_KEYS=y
```

Andy2244 commented:
* update to git (2019-02-27)
* adds "force group", "force user" options
* remove CONFIG_KEYS=y requirement
* switch to new repo location

PS: We worked out that the CONFIG_KEYS change is only needed for ACL, which is broken atm. So i could remove this requirement, so no worries about kernel changes anymore. It seems if we enable it for ACL the size will increase by 50kb, so maybe than we can create a module for it.

neheb commented:
I'm having a strange issue. On a little endian ramips device, cifsd works great.

On a big endian MIPS device, mount.cifs errors with cifs_mount failed w/return code = -5

Local to the router itself, it errors with:
```
mount: mounting //127.0.0.1/Music on test failed: I/O error
```

The setup uses the same smb.conf file, same cifsdb file, and same filesystem (ext4).

neheb commented:
Does this require the full libnl? I'm getting

```
[cifsd-worker/2888]: ERROR: Cannot resolve netlink family
```

spam on both ramips and ath79.

namjaejeon commented:
> On a big endian MIPS device, mount.cifs errors with cifs_mount failed w/return code = -5
Hm.. maybe, there is missing point to convert to little endian.
Could you please share dump log message after debug log is enable (echo 1 > /sys/module/cifs_server/parameters/cifsd_debugging) ?

And please move both issues to cifsd-team/ISSUE to check it with our members.
Thanks!

namjaejeon commented:
> Does this require the full libnl? I'm getting

Which libnl version are you using ?

neheb commented:
3.4.0 looks like.

Andy2244 commented:
* update to git (2019-03-05)

frymezim commented:
Tried and tested this and it's really fast compared to samba3, but my problem is that it doesn't show up on network list. I need to use the device IP for me to browse the drive.

```
[global]
        server string = NSA325
        workgroup = WORKGROUP
        netbios name = OpenWRT-NSA325
        map to guest = Bad User
        ipc timeout = 8
```

Andy2244 commented:
> Tried and tested this and it's really fast compared to samba3, but my problem is that it doesn't show up on network list. I need to use the device IP for me to browse the drive.

Yeah same problem as samba3/4, cifsd only has netbios support by itself. Windows10 does not show netbios shares by default anymore, its considered deprecated. We use wsdd2/avahi for samba4, but cifsd has no support for either. I have wsdd2 support on my to-do list, so this should add Windows10 support. I have no clue if and when cifsd will get avahi support for Linux/macOS.

bobafetthotmail commented:
I guess for the Linux/Mac side it should be possible to integrate it with umdns (the Avahi-like lightweight daemon of OpenWrt)
https://openwrt.org/docs/guide-developer/mdns

Like it was done for ssh https://github.com/openwrt/packages/commit/6cb6666c9783700613e700a4162cd09bf59daf9d

Andy2244 commented:
Sure, but umdns will only show the host machine, not the actual shares, so i'm not sure how usefully this is.

Andy2244 commented:
* update to git (2019-03-09)

PS: At Reviewers, is there any blocking issues left i should address for this PR?

neheb commented:
One blocking issue is that this is totally broken for MIPS(big endian) devices. I've been working with upstream to get the various issues fixed but it will take some time as cifsd has never been tested on big endian.

MIPS (big endian) is still the most common CPU type in OpenWrt.

Andy2244 commented:
@neheb Ah ok thanks for the help, as a ARM masterrace user i did not notice this.

neheb commented:
I am as well, but most of OpenWrt users run Archer C7v2s and other big endian MIPS devices.

neheb commented:
Problem seems fixed as of latest master.

namjaejeon commented:
@neheb Thanks a lot!

@Andy2244 I pushed all changes from cifsd-next to master branch.
You can use the latest master of cifsd.

Thanks!

Andy2244 commented:
* update to git (2019-03-13)
* fixes cifsd for MIPS (big endian) devices

@neheb Ok if you can confirm that this versions works on mips, i think we are gtg.

diizzyy commented:
Can we switch to codeload? :-)

Andy2244 commented:
> Can we switch to codeload? :-)

Sure + done, but it always confuses me what i need to change for codeload.
Could we maybe make this into a simple change like this, without the need to add custom PKG_SOURCE, PKG_BUILD_DIR?

`PKG_SOURCE_PROTO:=git` -> `PKG_SOURCE_PROTO:=git-codeload`

ynezz commented:
> Could we maybe make this into a simple PKG_SOURCE_PROTO change, without the need to add custom PKG_SOURCE, PKG_BUILD_DIR

Well, for example in your `cifsd` case, how would you then resolve complete URL, mainly the `cifsd-team` in the path?

> Ok if you can confirm that this versions works on mips, i think we are gtg.

What about the missing `procd` support?

BTW, I think, that having configuration handled via UCI would be nice to have as well, not just because of LuCI, but also for preserving configuration during sysupgrades, easier configuration etc.

Andy2244 commented:
> What about the missing `procd` support?

As noted, procd + cifsd was crashing/restarting my router. Is procd now required for all new packages?
I would rather spend my time on trying to get wsdd2 + luci/uci ready and put the procd issues on a lower priority.

ynezz commented:
> As noted, procd + cifsd was crashing/restarting my router. 

I know, I've also suggested to you to report it, so it could be fixed (From your description it looks like serious bug, so I guess, that it might have been fixed probably by now). The thing is, that some of the developers who might be able to fix this issue don't follow the GitHub parallel universe, but they do read bug reports occasionally.

> Is procd now required for all new packages?

Well I'm not aware about such rule, but I've just quickly checked daemons in the repository and can't find any example of package which has daemon and is not using `procd` for it's handling.

dedeckeh commented:
> > As noted, procd + cifsd was crashing/restarting my router.
> 
> I know, I've also suggested to you to report it, so it could be fixed (From your description it looks like serious bug, so I guess, that it might have been fixed probably by now). The thing is, that some of the developers who might be able to fix this issue don't follow the GitHub parallel universe, but they do read bug reports occasionally.
> 
> > Is procd now required for all new packages?
> 
> Well I'm not aware about such rule, but I've just quickly checked daemons in the repository and can't find any example of package which has daemon and is not using `procd` for it's handling.

There's very strong preference to use procd at least for the packages in the core repo

Andy2244 commented:
> I know, I've also suggested to you to report it, so it could be fixed (From your description it looks like serious bug, so I guess, that it might have been fixed probably by now).

Alright, i try to carve out some time next weekend to investigate this further and setup a proper bug report, only problem is Division 2 is out soon!

Andy2244 commented:
* update to git (2019-03-28)
* adds options (`bind interfaces only, interfaces, deadtime`)
* use procd (crashes router on stopping the service)

@namjaejeon Regarding the procd crash, maybe its because cifsd -n, will spawn a identically named `cifsd` watcher process? Can we rename this extra process or make it optional? What does it actually do?

sergey-senozhatsky commented:
> @namjaejeon Regarding the procd crash, maybe its because cifsd -n, will spawn a identically named `cifsd` watcher process? Can we rename this extra process or make it optional? 

I'm not Namjae, but we can rename that process, I guess...
I need to catch up with this whole discussion.

What does procd do if, say, firefox or chrome spawn a number of identically named processes?

> What does it actually do?

It does actual work (DCE RPC, IPC, etc.).

-ss

sergey-senozhatsky commented:
@Andy2244 does it make any difference if you run cifsd in "alternative" --no-detach mode `cifsd --n=2`?

Andy2244 commented:
> @Andy2244 does it make any difference if you run cifsd in "alternative" --no-detach mode `cifsd --n=2`?
Will try the new option, if i get back from a trip next week, thanks.

Andy2244 commented:
* updated to latest git (fixes procd crash)

@sergey-senozhatsky ok the latest update seem to have fixed the procd issue, this means `--n` now works, but `--n=2` still reboots/crashes. Does this make sense, what did you change for the `--n` option ?

sergey-senozhatsky commented:
On (04/09/19 04:09), Andy Walsh wrote:
>      • updated to latest git (fixes procd crash)
> 
>    [1]@sergey-senozhatsky ok the latest update seem to have fixed the procd
>    issue, this means --n now works, but --n=2 still reboots/crashes. Does
>    this make sense, what did you change for the --n option ?

Hmm. When was the last time you updated cifsd-tools? Did you test
previous cifsd-tools with commit 505ab8ef01c64248a8? (master branch)

We now setsid() in --n (--n=1) and don't setsid() in --n=2 mode.

	-ss

Andy2244 commented:
> We now setsid() in --n (--n=1) and don't setsid() in --n=2 mode.

So setsid() was not used before? My test build was a couple of weeks old.
As this seems to have fixed the procd issue, i can mention it for the procd issue. Not sure why procd crashes without it set.

thanks for the fix anyway

neheb commented:
@Andy2244 BUILD_DIR should probably be in the kernel directory.

This is not yet ready for pulling. Big endian issues strike again (I previously only tested mounting, not using).

Andy2244 commented:
> BUILD_DIR should probably be in the kernel directory.

Done

namjaejeon commented:
@Andy2244 Big endian issue is clear now. Rosen helped to improve the problem.

dengqf6 commented:
I've encountered `redefinition of 'put_unaligned_be64'` error when building against kernel 4.19
According to https://lore.kernel.org/patchwork/patch/729241/

I changed [this include](https://github.com/cifsd-team/cifsd/blob/c91c9b31773160a4d836f553789eb1dc9fc7f43f/unicode.c#L11) to `<asm/unaligned.h>` and it worked

sergey-senozhatsky commented:
@dengqf6 yeah, it seems that asm/unaligned.h is what we need to include.

`~/_next$ git grep "include <asm/unaligned.h>" | wc -l
522
`
Care to submit a patch (cifsd)?

dengqf6 commented:
Done
cifsd-team/cifsd#75

namjaejeon commented:
@dengqf6  Applied! Thanks for your work!

dengqf6 commented:
The latest git version builds flawlessly. But I cannot get it working on Windows 10.
It reports either `0x80070035` or `0x80070043` error, while not a single packet to TCP port 445 can be seen.
Is it a configuration issue, or it needs other packages such as wsdd2 to work?

neheb commented:
@dengqf6 should report upstream. Sounds like configuration issue.

namjaejeon commented:
@dengqf6 have you updated latest cifsd-tools as well kernel cifsd ? If you update only latest cifsd, It could not work.

Andy2244 commented:
* update to git (2019-04-21)

ynezz commented:
> see if anything else is missing for a initial commit?

I've just quickly skimmed through this PR again, and it still feels little bit alien to me without the UCI configuration, but I've already [expressed that](https://github.com/openwrt/openwrt/pull/1788#issuecomment-472395196).

Andy2244 commented:
> still feels little bit alien to me without the UCI configuration

Not sure why, nfs-kernel-server also has no extra UCI options. If that's what holding back this PR, i have to close it and let someone else take over. I don't have the time currently to create a full samba3/4 feature-set, including luci. So if anyone want to take cifsd feel free.

PS: Currently i'm busy switching samba4 to 4.10.x (python3), which has a nasty compile bug.

diizzyy commented:
I don't see why UCI would be a requirement

ynezz commented:
> including luci

UCI (OpenWrt's config system) is not LuCI (OpenWrt's web UI).

> I don't see why UCI would be a requirement

I'm not saying it's a requirement, I'm just trying to express, that using common OpenWrt infrastructure like procd and UCI are strong selling points, which would make me consider this PR as a candidate for merge. Until then I'm simply hesitant to spend (substantial amount of) time for build and run-time testing of this PR.

Please don't take me wrong, it's not NAK from me of any kind, anyone else could simply merge this as it is if it's good enough for them. I hope it's more clear now.

Andy2244 commented:
> UCI (OpenWrt's config system) is not LuCI (OpenWrt's web UI).

I know what UCI is, yet those two go hand in hand. Why would i create a UCI configuration that's not driven by a luci module? The idea for me is to make the package as easy to use as the samba package.

So the main reason to use UCI in this case, is mainly to-be able to setup shares + options via luci. Without luci, whats the point editing a UCI config, that is than used by the init script to feed the smb.conf, instead of directly editing the smb.conf?

ynezz commented:
> So the main reason to use UCI in this case, is to-be able to setup shares + options via luci. 

UCI is pretty much usable without LuCI as well. There's no LuCI in the OpenWrt core tree, so let's just focus on UCI now.

> Without luci, whats the point manually having to edit a UCI config, that is than used by the init script to feed the smb.conf, instead of directly editing the smb.conf?

UCI is pretty much usable without LuCI, I'm using it from C, shell and Lua pretty much for everything. The point of UCI is to have **U**nified **C**onfiguration **I**nterface instead of this fiddling directly with the config files, which is error prone, barely scriptable etc.

> PS: If i have more time i would like to bring cifsd on the same level as samba

No worries, I think, that there is no rush to get this merged ASAP, or is there?

Andy2244 commented:
> UCI is pretty much usable without LuCI, I'm using it from C, shell and Lua pretty much for everything.

I understand all this, its not that creating a luci/uci is really hard, i will pretty much reuse most of the samba stuff. The main problem is priorities and being able to test all this, i don't feel like working on the final uci/luci config stuff, before i can't actually make the shares show up correctly for a average user.

diizzyy commented:
Given that SMB 3.X has been EoL since years and 4.X is a no go on devices with limited storage space in general I can see why there's a need for something new/actively developed.
https://wiki.samba.org/index.php/Samba_Release_Planning

bobafetthotmail commented:
Eh, I'd say that merging this as-is (and adding a page about it in the wiki) would be enough to attract other (possibly low-skilled) contributors that can add the UCI configuration support, and raise awareness of this package/project.

Leaving it here as an unmerged PR will just make it bitrot.

neheb commented:
I agree

Andy2244 commented:
* cifsd: update to git (2019-05-09)

Andy2244 commented:
* cifsd/tools: update to git  (2019-05-19)
* adds UCI config options _(compatible with samba4 parameters)_
* add template logic _(manual created smb.conf will be kept)_

Example:
```
config globals
	option 'description'	'Cifsd on OpenWrt'

config share
	option name 'testshare'
	option path '/tmp'
	option guest_ok 'yes'
	option create_mask '0666'
	option dir_mask '0777'
	option read_only 'no'
	option force_root '1'
```
NOTE: I also process samba UCI section names, this should allow to quickly copy&paste a samba config into a cifsd.

PS: Hopefully now someone may pick this up and give it some test rounds, thanks!

neheb commented:
I'd run that init script through shellcheck.

neheb commented:
install: target '/home/mangix/devstuff/openwrt/build_dir/target-mipsel_24kc_musl/cifsd-tools-fef760c420dfb2a3622b9bd1e98ef46d48212282/.pkgdir/cifsd-tools/etc/cifsd/' is not a directory: No such file or directory

Andy2244 commented:
> I'd run that init script through shellcheck.
ok thanks, found and fixed a error.

> cifsd-tools/etc/cifsd/' is not a directory: No such file or directory
fixed, ofc a last minute manual error :)

Andy2244 commented:
* cifsd/tools: update to git (2019-05-31)

dengqf6 commented:
I think this should be moved to https://github.com/openwrt/packages
What do you think? @neheb

bobafetthotmail commented:
I think the idea is to replace the aging samba36 package with this.

samba36 is here (packages/network/services), not in packages.

In packages you can find samba4 package, which is much much larger and heavier on CPU

dengqf6 commented:
@bobafetthotmail cifsd-tools depends on glib2, which is in package repo

Andy2244 commented:
> cifsd-tools depends on glib2, which is in package repo

mhh that's a good point, i did overlook this dependencie. So i guess i need to move both packages to the package/feed?

bobafetthotmail commented:
Yes, that's where it must go. I don't think they will migrate glib2 into OpenWrt core repo, at least for now.

Best would be to drop the dependency on glib2, but I guess you cannot do that.

Andy2244 commented:
Ok new PR is over here openwrt/packages#9155

sergey-senozhatsky commented:
On (06/05/19 00:12), Alberto Bursi wrote:
>    Best would be to drop the dependency on glib2

Don't think this will happen in the foreseeable future, sorry.

	-ss
